### ISSUE TITLE
Is it possible to do a Retrofit call inside of an Interceptor?

### ISSUE THREAD

**Issue by mattblang** [square/okhttp#1351]
I am wanting to do a synchronized Retrofit call inside of the `intercept` function, but the OkHttp Dispatcher is killing the thread (I think).  Before I craft a request by hand I wanted to ask if I should be able to do this?  I am hitting a separate Retrofit service than the one this Interceptor is attached to, so I have no worry of infinite re-entry.  I also created an [SO question](http://stackoverflow.com/questions/28226765/make-a-synchronized-retrofit-call-from-inside-an-okhttp-interceptor), it contains my current Interceptor code.

**Comment by swankjesse** [comment]
I have a test case here:
https://github.com/square/okhttp/pull/1353

That uncovered a very significant bug, where our async requests weren't as concurrent as we thought they were. That said, that shouldn't impact your use of Retrofit.

Can you provide a stacktrace with the AssertionError? I'm puzzled at how you got there.

**Comment by mattblang** [comment]
@swankjesse There is no stacktrace actually.  If I put a debug point on the line with the synchronous Retrofit call, do a step-over, wait a few seconds (I can see in the OkHttp log that the synchronous call does run), then I arrive at that line in `OkHttp's Dispatcher`, followed by `Retrofit's ErrorHandler DEFAULT`, followed by `Retrofit's CallbackRunnable`, then the original request that had been intercepted completes.

**Comment by swankjesse** [comment]
Does this fix it?

```
    OkHttpClient okHttpClient = new OkHttpClient();
    okHttpClient.setDispatcher(new Dispatcher(new ThreadPoolExecutor(0, Integer.MAX_VALUE,
        60, TimeUnit.SECONDS, new SynchronousQueue<Runnable>())));

    RestAdapter restAdapter = new RestAdapter.Builder()
        .setEndpoint("http://" + server.getHostName() + ":" + server.getPort())
        .setClient(new OkClient(okHttpClient))
        .build();
```

**Comment by mattblang** [comment]
@swankjesse I added that to the OkHttpClient that the interceptor is attached to;  it didn't seem to change anything.

I want to mention that when the RetrofitError is thrown for "401 Unauthorized", it does contain the line where I do the synchronized call for a new token.  I didn't mention it earlier when we were talking about a stacktrace because this is an expected error if the token is expired.  Not sure if this is relevant at all, but I just wanted to mention.

`01-30 13:51:22.562    8120-8120/edu.msstate.nsparc.mdes D/﹕ 401 Unauthorized
    retrofit.RetrofitError: 401 Unauthorized
            at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:388)
            at retrofit.RestAdapter$RestHandler.invoke(RestAdapter.java:240)
            at java.lang.reflect.Proxy.invoke(Proxy.java:397)
            at $Proxy9.login(Unknown Source)
            **at edu.msstate.nsparc.mdes.retrofit.ReAuthInterceptor.intercept(ReAuthInterceptor.java:26)**
            at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:216)
            at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:192)
            at com.squareup.okhttp.Call.execute(Call.java:79)
            at retrofit.client.OkClient.execute(OkClient.java:53)
            at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:326)
            at retrofit.RestAdapter$RestHandler.access$100(RestAdapter.java:220)
            at retrofit.RestAdapter$RestHandler$2.obtainResponse(RestAdapter.java:278)
            at retrofit.CallbackRunnable.run(CallbackRunnable.java:42)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at retrofit.Platform$Android$2$1.run(Platform.java:142)
            at java.lang.Thread.run(Thread.java:818)`

**Comment by mattblang** [comment]
@swankjesse Also, at some point awhile later, if I am stepping over with the debugger then I make it to Retrofit's Platform.java at the end of the run block in the defaultHttpExecutor.  Again, not sure if this is relevant at all, but I want to include everything I can.

**Comment by swankjesse** [comment]
I'm unable to reproduce this. What's different?
https://gist.github.com/swankjesse/1e79c2558e13a061e07a

**Comment by mattblang** [comment]
@swankjesse The main difference is that I am making a call to `chain.proceed(request)`, checking the response code, then hitting a synchronized Retrofit call, all in the `interceptor`.  Does it still work if you make the `service.get()` call for `response2` within the interceptor?

Also, is it okay for me to reference `MyApplication.context` from the `interceptor`?

**Comment by swankjesse** [comment]
What do you mean, a _synchronized_ Retrofit call?

**Comment by mattblang** [comment]
@swankjesse Using the `Model functionName(@Body..)` call instead of the `void functionName(@Body, callback)` or `Observable<Model> functionName` call.

**Comment by JakeWharton** [comment]
Ah, that is synchronous not synchronized.

**Comment by mattblang** [comment]
@swankjesse @JakeWharton Ahh, my apologies.  Yes, that is a big difference in this context.

**Comment by swankjesse** [comment]
@mattblang could you provide an executable test case? I'm not able to reproduce.

**Comment by swankjesse** [comment]
No action to take yet, though I'm eager for a test case from @mattblang .

**Comment by mattblang** [comment]
@swankjesse My apologies, I had to go out of town this weekend.  I created a [test case](https://gist.github.com/mattblang/965121044a7065d32aca) similar to what I am doing at work.  It seems to work perfectly, so I need to figure out what variable is different.  I will check this first thing in the morning.

**Comment by swankjesse** [comment]
Thanks Matt! Let's figure this out.

**Comment by mattblang** [comment]
@swankjesse I figured out what was happening.  I wasn't catching a `RetrofitError` on the synchronous call.  I noticed the same behavior during my testing when a GSON parse exception occurred.  I then realized that our back-end returns `401 Unauthorized` if a non-existent web service is hit, so my guess is that I screwed up the route when I defined the synchronous call in the Service interface and was thrown off since 401 was what I wrote this Interceptor for.  Stupid mistake on my part, sorry for any time that was wasted.
